### PR TITLE
[JENKINS-73904] [JEP-237] check algorithm used for creating oic config is fips compliant

### PR DESCRIPTION
**Pull request by pankajy-dev**
https://issues.jenkins.io/browse/JENKINS-73904
https://github.com/jenkinsci/jep/tree/master/jep/237

### Submitter checklist
- [x] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x] Please describe what you did
- [x] Link to relevant issues in GitHub or Jira
- [x] Link to relevant pull requests, esp. upstream and downstream changes
- [x] Ensure you have provided tests - that demonstrates feature works or fixes the issue

### CODE REVIEW

File: src/main/java/org/jenkinsci/plugins/oic/OicAlgorithmValidator.java [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+
+    private static final boolean isFIPSMode = FIPS140.useCompliantAlgorithms();
+
+    /**
+     * Checks if the Jws signing algorithm used for OIC configuration is FIPS compliant.
+     */
+    public static boolean isJwsAlgorithmFipsNonCompliant(String algorithm) {
+        boolean matchNotFound = false;
+        if (isFIPSMode && algorithm != null) {
+            Set<JWSAlgorithm> jwsSupportedAlgorithms = new LinkedHashSet<>();
```
Comment (olamy) [inline src/main/java/org/jenkinsci/plugins/oic/OicAlgorithmValidator.java:36]: could be a static final field to avoid recreating the collection

File: src/main/java/org/jenkinsci/plugins/oic/OicAlgorithmValidator.java [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+        return matchNotFound;
+    }
+
+    /**
+     *  Checks if the Jwe encryption algorithm used for OIC configuration is FIPS compliant.
+     */
+    public static boolean isJweAlgorithmFipsNonCompliant(String algorithm) {
+        boolean matchNotFound = false;
+        if (isFIPSMode && algorithm != null) {
+            Set<JWEAlgorithm> jweSupportedAlgorithms = new LinkedHashSet<>();
```
Comment (olamy) [inline src/main/java/org/jenkinsci/plugins/oic/OicAlgorithmValidator.java:56]: same why not a static field with the collection already populated. I agree this will be helpful only if FIPS is activated :)

**Review by olamy — CHANGES_REQUESTED** [review]

File: src/main/java/org/jenkinsci/plugins/oic/OicSecurityRealm.java [hunk 3]
```diff
@@ -488,6 +488,13 @@ private OidcConfiguration buildOidcConfiguration() {
         // conf.setPreferredJwsAlgorithm(JWSAlgorithm.HS256);
         // set many more as needed...
 
+        // TODO if we set PreferredJwsAlgo then check for the validation
+        // if(conf.getPreferredJwsAlgorithm() != null && OicAlgorithmValidator.isAlgorithmNotFipsCompliant(
+        //                conf.getPreferredJwsAlgorithm().getName())){
+        //            throw new
+        // IllegalArgumentException(Messages.OicConfigNonCompliantAlgo_ErrorMessage(conf.getPreferredJwsAlgorithm().getName()));
+        //        }
+
```
Comment (PereBueno) [inline src/main/java/org/jenkinsci/plugins/oic/OicSecurityRealm.java:497]:
```suggestion
```

File: src/main/java/org/jenkinsci/plugins/oic/OicServerManualConfiguration.java [hunk 4]
```diff
@@ -156,6 +156,11 @@ public OIDCProviderMetadata toProviderMetadata() {
             ArrayList<JWSAlgorithm> allAlgorthms = new ArrayList<>();
             allAlgorthms.addAll(JWSAlgorithm.Family.HMAC_SHA);
             allAlgorthms.addAll(JWSAlgorithm.Family.SIGNATURE);
+
+            if (!allAlgorthms.isEmpty() && OicAlgorithmValidator.isJwsAlgoFipsNonCompliant(allAlgorthms)) {
+                throw new IllegalArgumentException(
```
Comment (PereBueno) [inline src/main/java/org/jenkinsci/plugins/oic/OicServerManualConfiguration.java:161]: Are we sure of this?
This will always launch an exception if there's a non compliant algorithm in the nimbus library (I guess there wil lalways be), and the user can not do anything to avoid it.
We should instead filter the list to allow only compliant algorithms

**Review by PereBueno — CHANGES_REQUESTED** [review]
We should be filtering existing algorithms, not launching an exception

**Review by PereBueno — CHANGES_REQUESTED** [review]
Don't know what happened, previous comment was a change request, not a comment

**Comment by pankajy-dev** [comment]
> Don't know what happened, previous comment was a change request, not a comment

Just realised I got review comments on this PR. Some of these reviews were already provided by James on the PR that I created on his Fork.

I addressed few of them.

File: src/main/java/org/jenkinsci/plugins/oic/OicAlgorithmValidator.java (hunk 1 above)
Comment (pankajy-dev) [reply src/main/java/org/jenkinsci/plugins/oic/OicAlgorithmValidator.java]: Done

File: src/main/java/org/jenkinsci/plugins/oic/OicAlgorithmValidator.java (hunk 2 above)
Comment (pankajy-dev) [reply src/main/java/org/jenkinsci/plugins/oic/OicAlgorithmValidator.java]: Done

File: src/main/java/org/jenkinsci/plugins/oic/OicSecurityRealm.java (hunk 3 above)
Comment (pankajy-dev) [reply src/main/java/org/jenkinsci/plugins/oic/OicSecurityRealm.java]: Removed this uncommented code, as no algorithm is used here.

**Comment by pankajy-dev** [comment]
Not able to add reviewers here.
@jtnord , @PereBueno , @olamy , @fcojfernandez  - Please review.

File: src/main/java/org/jenkinsci/plugins/oic/OicAlgorithmValidator.java [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+        filterFipsNonCompliantAlgorithms(algorithms, OicAlgorithmValidator::isJwsAlgorithmFipsNonCompliant);
+    }
+
+    /**
+     *  Checks if the encryption method used for OIC configuration is FIPS compliant.
+     */
+    public static boolean isEncryptionMethodFipsNonCompliant(Algorithm encryptionMethod) {
+        boolean matchNotFound = false;
+        if (isFIPSMode && encryptionMethod != null) {
+            if (!supportedEncryptionMethod.isEmpty()) {
```
Comment (PereBueno) [inline src/main/java/org/jenkinsci/plugins/oic/OicAlgorithmValidator.java:110]: Can it be empty? IINW it's final and filled in the static block

File: src/main/java/org/jenkinsci/plugins/oic/OicAlgorithmValidator.java [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+import java.util.List;
+import java.util.Set;
+import java.util.function.Function;
+import jenkins.security.FIPS140;
+
+/**
+ * This class helps in validating algorithms for FIPS compliance and filtering the non-compliant algorithms when in
+ * FIPS mode.
+ */
+public class OicAlgorithmValidator {
```
Comment (fcojfernandez) [inline src/main/java/org/jenkinsci/plugins/oic/OicAlgorithmValidator.java:26]: This is only for FIPS
```suggestion
/**
 * This class helps in validating algorithms for FIPS-140 compliance and filtering the non-compliant algorithms when in
 * FIPS mode.
 */
public class OicAlgorithmValidatorFIPS140 {
```

File: src/main/java/org/jenkinsci/plugins/oic/OicServerManualConfiguration.java [hunk 7]
```diff
@@ -155,7 +155,8 @@ public OIDCProviderMetadata toProviderMetadata() {
             // rather we just say "I support anything, and let the check for the specific ones fail and fall through
             ArrayList<JWSAlgorithm> allAlgorthms = new ArrayList<>();
             allAlgorthms.addAll(JWSAlgorithm.Family.HMAC_SHA);
-            allAlgorthms.addAll(JWSAlgorithm.Family.SIGNATURE);
+            allAlgorthms.addAll(JWSAlgorithm.Family.RSA);
+            allAlgorthms.addAll(JWSAlgorithm.Family.EC);
```
Comment (PereBueno) [inline src/main/java/org/jenkinsci/plugins/oic/OicServerManualConfiguration.java:159]: Shouldn't we be filtering here `allAlgorithms` here?

File: src/main/java/org/jenkinsci/plugins/oic/OicAlgorithmValidator.java
Comment (fcojfernandez) [inline src/main/java/org/jenkinsci/plugins/oic/OicAlgorithmValidator.java:1]: This class is exposing methods to check FIPS-140. If those methods were private we could discuss if having a negated output is correct or not, but in a utility class like this, checking `sometingNon` makes the code hard to follow and maintain

Comment (fcojfernandez) [reply src/main/java/org/jenkinsci/plugins/oic/OicAlgorithmValidator.java:1]: Besides, modifying the lists received as parameter would be risky. It's better to always return an immutable list

File: src/main/java/org/jenkinsci/plugins/oic/OicSecurityRealm.java [hunk 8]
```diff
@@ -509,6 +513,107 @@ private OidcConfiguration buildOidcConfiguration() {
         return conf;
     }
 
+    private static void filterJwsAlgorithms(OIDCProviderMetadata oidcProviderMetadata) {
```
Comment (PereBueno) [inline src/main/java/org/jenkinsci/plugins/oic/OicSecurityRealm.java:516]: Looks dangerous, what if we're using an immutableList? All the `set*JW*Algs` methods are public

File: src/main/java/org/jenkinsci/plugins/oic/OicAlgorithmValidator.java [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+        jweSupportedAlgorithms.addAll(PasswordBasedCryptoProvider.SUPPORTED_ALGORITHMS);
+
+        // Init complaint EncryptionMethods
+        supportedEncryptionMethod.addAll(ContentCryptoProvider.SUPPORTED_ENCRYPTION_METHODS);
+    }
+
+    /**
+     * Checks if the Jws signing algorithm used for OIC configuration is FIPS compliant.
+     */
+    public static boolean isJwsAlgorithmFipsNonCompliant(Algorithm algorithm) {
```
Comment (PereBueno) [inline src/main/java/org/jenkinsci/plugins/oic/OicAlgorithmValidator.java:55]: As a rule of thumb, "positive" method (e.g.: `isJwsAlgorithmCompliant`) are easier to follow (those double negations make this difficult to read)

File: src/test/java/org/jenkinsci/plugins/oic/OicSecurityRealmNonFIPSAlgoTest.java
Comment (PereBueno) [inline src/test/java/org/jenkinsci/plugins/oic/OicSecurityRealmNonFIPSAlgoTest.java:1]: Test name is wrong, we're testing FIPS `true` here

File: src/test/java/org/jenkinsci/plugins/oic/OicSecurityRealmFIPSAlgoTest.java
Comment (PereBueno) [inline src/test/java/org/jenkinsci/plugins/oic/OicSecurityRealmFIPSAlgoTest.java:1]: Same, wrong name

File: src/test/java/org/jenkinsci/plugins/oic/OicSecurityRealmFIPSAlgoTest.java [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+    public void doCheckAlgorithmFilter() throws IOException, ParseException {
+        OIDCProviderMetadata oidcProviderMetadata = OicSecurityRealmNonFIPSAlgoTest.getNonCompliantMockObject();
+        int countIDTokenJWSAlgs = oidcProviderMetadata.getIDTokenJWSAlgs().size();
+        int countIDTokenJWEAlgs = oidcProviderMetadata.getIDTokenJWEAlgs().size();
+        OicSecurityRealm.filterNonCompliantAlgorithms(oidcProviderMetadata);
+        assertEquals(
+                countIDTokenJWSAlgs, oidcProviderMetadata.getIDTokenJWSAlgs().size());
+        assertEquals(
+                countIDTokenJWEAlgs, oidcProviderMetadata.getIDTokenJWEAlgs().size());
+    }
```
Comment (PereBueno) [inline src/test/java/org/jenkinsci/plugins/oic/OicSecurityRealmFIPSAlgoTest.java:40]: Filtering isn't right, adding just the opposite check to this test fails
```
    @Test
    public void doCheckAlgorithmFilter() throws IOException, ParseException {
        OIDCProviderMetadata oidcProviderMetadata = OicSecurityRealmNonFIPSAlgoTest.getNonCompliantMockObject();
        int countIDTokenJWSAlgs = oidcProviderMetadata.getIDTokenJWSAlgs().size();
        int countIDTokenJWEAlgs = oidcProviderMetadata.getIDTokenJWEAlgs().size();
        OicSecurityRealm.filterNonCompliantAlgorithms(oidcProviderMetadata);
        assertEquals(
                countIDTokenJWSAlgs, oidcProviderMetadata.getIDTokenJWSAlgs().size());
        assertEquals(
                countIDTokenJWEAlgs, oidcProviderMetadata.getIDTokenJWEAlgs().size());

        // reset the mock, in FIPs mode now
        fips140Mock.reset();
        fips140Mock.when(FIPS140::useCompliantAlgorithms).thenReturn(true);
        // restart the object
        oidcProviderMetadata = OicSecurityRealmNonFIPSAlgoTest.getNonCompliantMockObject();
        // filter
        OicSecurityRealm.filterNonCompliantAlgorithms(oidcProviderMetadata);
        // We should not have the same number of algorithms as in non FIPS mode
        assertNotEquals(countIDTokenJWSAlgs, oidcProviderMetadata.getIDTokenJWSAlgs().size());
        assertNotEquals(countIDTokenJWEAlgs, oidcProviderMetadata.getIDTokenJWEAlgs().size());
    }
```

**Review by PereBueno — CHANGES_REQUESTED** [review]

File: src/main/java/org/jenkinsci/plugins/oic/OicServerManualConfiguration.java (hunk 7 above)
Comment (pankajy-dev) [reply src/main/java/org/jenkinsci/plugins/oic/OicServerManualConfiguration.java:159]: these are the compliant algorithms, thats why haven't put filter here.

File: src/main/java/org/jenkinsci/plugins/oic/OicAlgorithmValidator.java (hunk 5 above)
Comment (pankajy-dev) [reply src/main/java/org/jenkinsci/plugins/oic/OicAlgorithmValidator.java]: Done

File: src/main/java/org/jenkinsci/plugins/oic/OicAlgorithmValidator.java [hunk 11]
```diff
@@ -0,0 +1,145 @@
...
+import java.util.List;
+import java.util.Set;
+import java.util.function.Function;
+import jenkins.security.FIPS140;
+
+/**
+ * This class helps in validating algorithms for FIPS-140 compliance and filtering the non-compliant algorithms when in
+ * FIPS mode.
+ */
+public class OicAlgorithmValidatorFIPS140 {
```
Comment (PereBueno) [inline src/main/java/org/jenkinsci/plugins/oic/OicAlgorithmValidator.java:26]: Class name must match file name

**Review by PereBueno — CHANGES_REQUESTED** [review]

File: src/main/java/org/jenkinsci/plugins/oic/OicAlgorithmValidator.java (hunk 9 above)
Comment (pankajy-dev) [reply src/main/java/org/jenkinsci/plugins/oic/OicAlgorithmValidator.java:57]: Done

File: src/test/java/org/jenkinsci/plugins/oic/OicSecurityRealmNonFIPSAlgoTest.java
Comment (pankajy-dev) [reply src/test/java/org/jenkinsci/plugins/oic/OicSecurityRealmNonFIPSAlgoTest.java:1]: Done

File: src/main/java/org/jenkinsci/plugins/oic/OicServerManualConfiguration.java (hunk 7 above)
Comment (fcojfernandez) [reply src/main/java/org/jenkinsci/plugins/oic/OicServerManualConfiguration.java:159]: wrong filter

File: src/main/resources/org/jenkinsci/plugins/oic/Messages.properties [hunk 12]
```diff
@@ -25,3 +25,5 @@ OicSecurityRealm.TokenRequestFailure = Token request failed: {0}"
 OicSecurityRealm.TokenRefreshFailure = Unable to refresh access token
 OicServerWellKnownConfiguration.DisplayName = Discovery via well-known endpoint
 OicServerManualConfiguration.DisplayName = Manual entry
+OicConfigNonCompliantAlgo.ErrorMessage = Algorithm ''{0}'' used for OIC configuration is not FIPS compliant.
```
Comment (fcojfernandez) [inline src/main/resources/org/jenkinsci/plugins/oic/Messages.properties:28]: unused

File: src/main/java/org/jenkinsci/plugins/oic/OicSecurityRealm.java [hunk 13]
```diff
@@ -509,6 +513,148 @@ private OidcConfiguration buildOidcConfiguration() {
...
```
Comment (jtnord) [inline src/main/java/org/jenkinsci/plugins/oic/OicSecurityRealm.java:516]: NIT: this class is already getting quire big and as this is static these methods could be moved to a utility class.

File: src/main/java/org/jenkinsci/plugins/oic/OicAlgorithmValidatorFIPS140.java [hunk 14]
```diff
@@ -0,0 +1,129 @@
...
+        jweSupportedAlgorithms.addAll(PasswordBasedCryptoProvider.SUPPORTED_ALGORITHMS);
+
+        // Init complaint EncryptionMethods
+        supportedEncryptionMethod.addAll(ContentCryptoProvider.SUPPORTED_ENCRYPTION_METHODS);
+        // XC20P is not complaint method
+        supportedEncryptionMethod.remove(EncryptionMethod.XC20P);
+    }
+
+    /**
+     * Checks if the Jws signing algorithm used for OIC configuration is FIPS-140 compliant.
```
Comment (jtnord) [inline src/main/java/org/jenkinsci/plugins/oic/OicAlgorithmValidatorFIPS140.java:55]: although it doesn't.
as this returns `true` for something that is not compliant when not in FIPS mode.
Would it be better to just not perform the call to do any filtering if we are not in FIPS mode and then these methods would do what they are documented to do and what is expected by the function names?

File: src/main/java/org/jenkinsci/plugins/oic/OicSecurityRealm.java [hunk 15]
```diff
@@ -509,6 +513,148 @@ private OidcConfiguration buildOidcConfiguration() {
...
+        // Filter Jws Algorithms
+        if (oidcProviderMetadata.getIDTokenJWSAlgs() != null) {
```
Comment (jtnord) [inline src/main/java/org/jenkinsci/plugins/oic/OicSecurityRealm.java:518]: remove the `null` checks and make the filter handle `null` input?

e.g by handling the `null` in `filterFipsNonCompliantJwsAlgorithm` you can have

`oidcProviderMetadate.setIdTokenJWSAlgs(OicAlgorithmValidatorFIPS140.filterFipsNonCompliantJwsAlgorithm(oidcProviderMetadata.getIDTokenJWSAlgs());`

File: src/main/java/org/jenkinsci/plugins/oic/OicAlgorithmValidatorFIPS140.java [hunk 16]
```diff
@@ -0,0 +1,129 @@
...
+        jwsSupportedAlgorithms.addAll(MACSigner.SUPPORTED_ALGORITHMS);
+        jwsSupportedAlgorithms.addAll(RSASSASigner.SUPPORTED_ALGORITHMS);
+        jwsSupportedAlgorithms.addAll(ECDSASigner.SUPPORTED_ALGORITHMS);
+
+        // Init compliant Jwe algorithms
+        jweSupportedAlgorithms.addAll(AESCryptoProvider.SUPPORTED_ALGORITHMS);
+        jweSupportedAlgorithms.addAll(RSACryptoProvider.SUPPORTED_ALGORITHMS);
+        // RSA1_5 is deprecated and not a compliant algorithm.
+        jweSupportedAlgorithms.remove(JWEAlgorithm.RSA1_5);
+        jweSupportedAlgorithms.addAll(ECDHCryptoProvider.SUPPORTED_ALGORITHMS);
```
Comment (jtnord) [inline src/main/java/org/jenkinsci/plugins/oic/OicAlgorithmValidatorFIPS140.java:45]: seems to use [its own key derivation function](https://bitbucket.org/connect2id/nimbus-jose-jwt/src/94a2e4d11ce322966d30a551a2d482936b588c87/src/main/java/com/nimbusds/jose/crypto/impl/ECDHCryptoProvider.java?at=master#lines-104:139) ([ConcatKDF](https://bitbucket.org/connect2id/nimbus-jose-jwt/src/94a2e4d11ce322966d30a551a2d482936b588c87/src/main/java/com/nimbusds/jose/crypto/impl/ConcatKDF.java?at=master#ConcatKDF.java-48,70)) and so would not be compliant?

**Review by jtnord — CHANGES_REQUESTED** [review]
to me it looks like none of the the `ECDHCryptoProviders` would be compliant?

**Comment by jtnord** [comment]
Subsumed in #428
